src/main/java/com/aws/greengrass/ggq/Exec.java

```diff
@@ -0,0 +1,466 @@
+    void setClosed() {
+        if (!isClosed.get()) {
+            final IntConsumer wd = whenDone;
+            final int exit = process == null ? -1 : process.exitValue();
+            isClosed.set(true);
+            if (wd != null) {
+                wd.accept(exit);
+            }
+        }
+    }
+
+    public boolean isRunning() {
+        return process == null ? !isClosed.get() : process.isAlive();
+    }
+
+    /**
+     * Get associated process instance representing underlying OS process.
+     *
+     * @return process object.
+     */
+    public Process getProcess() {
+        return process;
+    }
+
+    @Override
+    public synchronized void close() throws IOException {
+        Process p = process;
+        if (p != null && p.isAlive()) p.destroyForcibly();
+    }
+
+    @Override
+    public String toString() {
+        return Utils.deepToString(cmds, 90).toString();
+    }
+
+    /**
+     * Sends the lines of an InputStream to a consumer in the background.
+     */
+    private class Copier extends Thread {
+        private final Consumer<CharSequence> out;
+        private final InputStream in;
+        private int nlines = 0;
+
+        Copier(InputStream i, Consumer<CharSequence> s) {
+            super("Copier");
+            in = i;
+            out = s;
+            // Set as a daemon thread so that it dies when the main thread exits
+            setDaemon(true);
+            if (whenDone != null) {
+                if (numberOfCopiers == null) {
+                    numberOfCopiers = new AtomicInteger(1);
+                } else {
+                    numberOfCopiers.incrementAndGet();
+                }
+            }
+        }
+        
+        public int getNlines() { return nlines; }
+
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8), 200)) {
+                StringBuilder sb = new StringBuilder();
+                while (true) {
+                    int c;
+                    for (c = br.read(); c >= 0 && c != '\n'; c = br.read()) {
+                        sb.append((char) c);
+                    }
+                    if (c >= 0) {
+                        sb.append('\n');
+                        nlines++;
+                    }
+                    if (out != null && sb.length() > 0) {
+                        out.accept(sb);
+                    }
+                    sb.setLength(0);
+                    if (c < 0) {
+                        break;
+                    }
+                }
+            } catch (Throwable ignore) {
+                // nothing that can go wrong here worries us, they're
+                // all EOFs
+            }
+            if (whenDone != null && numberOfCopiers.decrementAndGet() <= 0) {
+                try {
+                    process.waitFor();
+                    setClosed();
+                } catch (InterruptedException ignore) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)